ambari-server/src/main/java/org/apache/ambari/server/controller/AddServiceRequest.java

```diff
@@ -0,0 +1,252 @@
+import org.apache.ambari.server.controller.internal.ProvisionAction;
+import org.apache.ambari.server.topology.ConfigRecommendationStrategy;
+import org.apache.ambari.server.topology.Configurable;
+import org.apache.ambari.server.topology.Configuration;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import io.swagger.annotations.ApiModel;
+import io.swagger.annotations.ApiModelProperty;
+
+/**
+ * Data object representing an add service request.
+ */
+@ApiModel
+@JsonInclude(JsonInclude.Include.NON_EMPTY)
+public final class AddServiceRequest {
+
+  static final String OPEATION_TYPE = "operation_type";
+  static final String CONFIG_RECOMMENDATION_STRATEGY = "config_recommendation_strategy";
+  static final String PROVISION_ACTION = "provision_action";
+  static final String STACK_NAME = "stack_name";
+  static final String STACK_VERSION = "stack_version";
+  static final String SERVICES = "services";
+  static final String COMPONENTS = "components";
+
+  private final OperationType operationType;
+  private final ConfigRecommendationStrategy recommendationStrategy;
+  private final ProvisionAction provisionAction;
+  private final String stackName;
+  private final String stackVersion;
+  private final Set<Service> services;
+  private final Set<Component> components;
+  private final Configuration configuration;
+
+  @JsonCreator
+  public AddServiceRequest(@JsonProperty(value = OPEATION_TYPE, required = true) OperationType operationType,
+                           @JsonProperty(CONFIG_RECOMMENDATION_STRATEGY) ConfigRecommendationStrategy recommendationStrategy,
+                           @JsonProperty(PROVISION_ACTION)ProvisionAction provisionAction,
+                           @JsonProperty(STACK_NAME) String stackName,
+                           @JsonProperty(STACK_VERSION) String stackVersion,
+                           @JsonProperty(SERVICES) Set<Service> services,
+                           @JsonProperty(COMPONENTS)Set<Component> components,
+                           @JsonProperty(Configurable.CONFIGURATIONS) Collection<? extends Map<String, ?>> configs) {
+    this(operationType, recommendationStrategy, provisionAction, stackName, stackVersion, services, components,
+      Configurable.parseConfigs(configs));
+  }
+
+
+  private AddServiceRequest(OperationType operationType,
+                            ConfigRecommendationStrategy recommendationStrategy,
+                            ProvisionAction provisionAction,
+                            String stackName,
+                            String stackVersion,
+                            Set<Service> services,
+                            Set<Component> components,
+                            Configuration configuration) {
+    this.operationType = operationType;
```

Comment:
Minor issue: We should probably add a similar check for null, and default to "ADD_SERVICE", if the operation type is not specified in the request.

Reply:
Fixed